benchmark/C/Savina/src/concurrency/BoundedBuffer.lf

```diff
@@ -0,0 +1,230 @@
+ * this benchmark, all producers, consumers and the manager operate fully
+ * asynchronously. However, in the LF implementation everything is synchronous.
+ * Thus, strictly speaking, the buffer isn't actually needed as we can design
+ * the program so that every tick a producer produces data and a consumer
+ * consumes it. But, the number of consumers and producers can be selected
+ * arbitrarily. Therefore, we still need some logic that takes care of
+ * orchestrating the producers according to the rate at which the consumers
+ * consume. Also, we could think about using an alternative design, where the
+ * consumers don't become available again immediately after consuming (at a
+ * later tag). Therefore, this benchmark still implements most of
+ * synchronization logic of the original manager in the Akka version.
+ * 
+ * @author Christian Menard
+ * @author Hannes Klein
+ * @author Matt Chorlian
+ */
+
+target C {
+    /* [[[cog
+      if (threaded_runtime=="True"):
+          cog.outl(f"threads: {threads},")
+      else:
+          cog.outl("threads: 0,")
+    ]]] */
+    threads: 0,
+    /// [[[end]]]
+    files: ["/lib/c/reactor-c/util/deque.h", "/lib/c/reactor-c/util/deque.c", "../include/PseudoRandom.h"],
+    flags: "-lm"
+};
+
+preamble {=
+    #include "PseudoRandom.h"
+    #include <math.h>
+    
+    double processItem(double curTerm, size_t cost) {
+        double res = curTerm;
+        
+        //FIXME: segfaulting here
+        PseudoRandom random = malloc(sizeof(struct PseudoRandom));
+        random->mValue = malloc(sizeof(long));
+        random->mValue = cost;
+
+        if(cost > 0) {
+            for(size_t i = 0; i < cost; i++) {
+                for(size_t j = 0; j < 100; j++) {
+                    res += log(labs(nextDouble(random)) + 0.01);
+                }
+            }
+        } else {
+            res += log(labs(nextDouble(random)) + 0.01);
+        }
+        
+        //free
+        free(random);
+        free(random->mValue);
+        return res;
+    }
+=}
+
+
+
+
+reactor ManagerReactor(bufferSize: size_t(50), numProducers: size_t(40), numConsumers: size_t(40)) {
+
+    preamble {=
+        #include "deque.c"
+    =}
+    
+    //FIXME: need to cog this
+    state adjustedBuffersize: size_t(10);
+    state pendingData: deque_t;
+    state numTerminatedProducers: size_t(0);
+    state producerTerminated: {=bool[]=};
+    state producer_id: size_t(0);
+    
+    input start: bool;
+    output finished: bool;
+    
+    input[numProducers] producerData: double;
+    input[numProducers] producerFinished: bool;
+    output[numProducers] producerCommand: bool;
+    
+    input[numConsumers] consumerAvailable: bool;
+    output[numConsumers] consumerData: double;
+    
+    reaction(startup) -> producerCommand {=
+        // reset local state
+        self->numTerminatedProducers = 0;
+        self->producerTerminated[self->numProducers];
```

Comment:
It doesn't work like that in C, unfortunately. The array needs to be allocated here